### PR TITLE
Issue 103 faster csv

### LINKED ISSUE #103: Runtime Improvement of CSVWriter
Currently, the csv writer is really slow, would be nice if it could be faster. perhaps through parallelised write out or at least no waiting for the writer.

### PR DESCRIPTION
**Pull request by ChristianVonMach**
runtime improved version of csvwriter. should close #103

### CODE REVIEW

**Review by LinhWuerzburger — CHANGES_REQUESTED** [review]
README has to be updated. ("if you already have a local copy of ARTSS and are missing spdlog do a recursive submodule init.")

File: src/visualisation/CSVWriter.cpp [hunk 1]
```diff
@@ -4,132 +4,72 @@
...
-    delete[] (v_vel);
-    delete[] (w_vel);
-    delete[] (pres);
-    delete[] (Temp);
 }
 
-void CSVWriter::csv_write(const char *filename, real **vars, int size_vars, const char **var_names) {
+void CSVWriter::csv_write(const std::string& filename, real **vars, int size_vars,
+        const std::vector<std::string>& var_names) {
+#ifndef BENCHMARKING
```
Comment (LinhWuerzburger) [inline src/visualisation/CSVWriter.cpp:72]: already in time integration. not necessary and in my opinion also not the right place.

File: src/visualisation/CSVWriter.cpp [hunk 2]
```diff
@@ -142,65 +82,52 @@ void CSVWriter::csv_write(const char *filename, real **vars, int size_vars, cons
...
-                for (auto & coord : coords) {
-                    outputFile << coord[index] << delimiter;
-                }
+                result = fmt::format(FMT_COMPILE("{},{},{},{},"),
+                                                i, j, k, index);
+
+                result.append(fmt::format(FMT_COMPILE("{},{},{},"),
+                                                coords[0][index],
+                                                coords[1][index],
+                                                coords[2][index]));
```
Comment (LinhWuerzburger) [inline src/visualisation/CSVWriter.cpp:114]: why not use i, j and k ?

File: src/visualisation/CSVWriter.cpp [hunk 3]
```diff
@@ -142,65 +82,52 @@ void CSVWriter::csv_write(const char *filename, real **vars, int size_vars, cons
     real dy = domain->get_dy();
     real dz = domain->get_dz();
 
-    const char *coord_names[] = {"i", "j", "k", "index", "x-coords (m)", "y-coords (m)", "z-coords (m)"};
+    std::vector<std::string> coord_names = {"i", "j", "k", "index",
+                                            "x-coords (m)", "y-coords (m)", "z-coords (m)"};
 
     auto *x_centres = new float[size];
     auto *y_centres = new float[size];
     auto *z_centres = new float[size];
```
Comment (LinhWuerzburger) [inline src/visualisation/CSVWriter.cpp:90]: we should change the centre values to real as well, for consistency reason. Guess some old code which wasn't changed yet

File: src/visualisation/CSVWriter.cpp (hunk 2 above)
Comment (LinhWuerzburger) [reply src/visualisation/CSVWriter.cpp:114]: I missread. Going to close this

**Comment by LinhWuerzburger** [comment]
In addition, I prefer the style where you emphasise the type of the pointed-to data. `someType *somePtr` rather than `someType* somePtr`. My reasoning is mainly that the pointer belongs to the data type because if you define multiple pointers in one line, you have to write: `someType *p1, *p2` while `someType* p1, p2` creates only one pointer.

We should add this at some point to our guide lines I guess... the same with underscores....

**Comment by ChristianVonMach** [comment]
> README has to be updated. ("if you already have a local copy of ARTSS and are missing spdlog do a recursive submodule init.")

done in 48fc47d. Added fmt and spdlog as dependency.

File: src/visualisation/CSVWriter.cpp (hunk 3 above)
Comment (ChristianVonMach) [reply src/visualisation/CSVWriter.cpp]: updatet in 70b52fc

**Comment by ChristianVonMach** [comment]
> In addition, I prefer the style where you emphasise the type of the pointed-to data. `someType *somePtr` rather than `someType* somePtr`. My reasoning is mainly that the pointer belongs to the data type because if you define multiple pointers in one line, you have to write: `someType *p1, *p2` while `someType* p1, p2` creates only one pointer.
> 
> We should add this at some point to our guide lines I guess... the same with underscores....

done in 4275acd

File: src/visualisation/CSVWriter.cpp (hunk 1 above)
Comment (LinhWuerzburger) [reply src/visualisation/CSVWriter.cpp:72]: cannot be resolved without more effort. therefore it won't be changed now

File: src/TimeIntegration.cpp [hunk 4]
```diff
@@ -119,7 +121,11 @@ void TimeIntegration::run() {
 #pragma acc update host(d_nu_t[:bsize])
 #pragma acc update host(d_S_T[:bsize]) wait    // all in one update does not work!
 
-            m_visual->visualise(m_field_controller, t_cur);
+            if (m_has_analytical_solution) {
+                m_solution->calc_analytical_solution(t_cur);
+            }
```
Comment (LinhWuerzburger) [inline src/TimeIntegration.cpp:126]: this inquiry does not belong to the TimeIntegration class. It can be in the Visual class as before or in the Solution class itself. Either way, this shouldn't be part of this PR.

**Review by LinhWuerzburger — CHANGES_REQUESTED** [review]

File: src/TimeIntegration.cpp (hunk 4 above)
Comment (ChristianVonMach) [reply src/TimeIntegration.cpp]: since `Solution::calc_analytical_solution` is not const but `Visual::visualise` is now we had to move it. The only option left is `Solution`. Done in e42733f

File: src/analysis/Solution.cpp [hunk 5]
```diff
@@ -129,6 +132,11 @@ void Solution::calc_analytical_solution(real t) {
     if (m_current_time_step == t) {
         return;
     }
+
+    if (!m_has_analytical_solution) {
```
Comment (LinhWuerzburger) [inline src/analysis/Solution.cpp:136]: should we combine it with the inquiry before?

**Review by LinhWuerzburger — CHANGES_REQUESTED** [review]

File: src/TimeIntegration.cpp [hunk 6]
```diff
@@ -31,9 +31,11 @@ TimeIntegration::TimeIntegration(SolverController *sc) {
 
     m_adaption = new Adaption(m_field_controller);
 #ifndef BENCHMARKING
+    m_has_analytical_solution = (params->get("solver/solution/available") == "Yes");
```
[the remark this answers is not in the document]
Comment (ChristianVonMach) [reply src/TimeIntegration.cpp]: indeed. Cleaned up in dfcc5e5

File: src/analysis/Solution.cpp (hunk 5 above)
Comment (ChristianVonMach) [reply src/analysis/Solution.cpp:136]: i dont think so. Each query has its own semantic purpose. The first one checking it the simulation finished the second if an analytical solution exists. The compiler should optimize this if it makes a difference. I my opinion this is not worth mixing up code.

Comment (LinhWuerzburger) [reply src/analysis/Solution.cpp:136]: ok

**Review by LinhWuerzburger — APPROVED** [review]
